File: src/depsolver/solver/propositions/validity.py
"""
All states are required to fulfil the dependency and conflict constraints.
"""
from .utils import disjunction, conjunction
from z3 import And, Not, Implies
from functools import partial

def make_validity_constraint(package_list,
                             package_variables,
                             step_limit,
                             expand_version_constraint):
    validity_constraints = [
        package_validity_constraint(package,
                                    package_variables[time][package.id],
                                    partial(expand_version_constraint, time))
        for package in package_list
        for time in range(step_limit)]
    return conjunction(validity_constraints)

def package_validity_constraint(package,
                                package_variable,
                                version_constraint_expander):
    make_dependency_constraint = \
        partial(dependency_constraint,
                package_variable,
                version_constraint_expander)
    make_conflict_constraint = \
        partial(conflict_constraint,
                package_variable,
                version_constraint_expander)

    all_dependencies_installed = \
        conjunction(map(make_dependency_constraint, package.dependencies)) \
        if package.dependencies else True

    no_conflicting_installed = \
        conjunction(map(make_conflict_constraint, package.conflicts)) \
        if package.conflicts else True

    return And(all_dependencies_installed, no_conflicting_installed)

def conflict_constraint(package_variable,
                        matching,
                        conflict):
    #  ¬(package_variable and (or... matching(conflict)))
    # TODO This is slightly broken for cases where constraint does not match
    # any package.
    matches = list(matching(conflict))
    return Not(And(package_variable, disjunction(matches))) \
        if matches else True

def dependency_constraint(package_variable,
                          matching,
                          dependency):
    def alternative_constraint(alternative):
        # TODO This is broken for cases where constraint does not match
        # any package. If this is the case we should probably error.
        alternative_as_list = list(matching(alternative))
        return disjunction(alternative_as_list) \
            if alternative_as_list else True


    #  TODO This is broken for empty list of alternatives.
    any_of_alternatives = list(map(alternative_constraint, dependency))
    return Implies(package_variable, disjunction(any_of_alternatives)) \
        if any_of_alternatives else True

File: rust/src/depsolver/solver/propositions/validity.rs
use crate::depsolver::model::constraint::PackageConstraint;
use crate::depsolver::model::package::Package;
use z3::Ast;
use super::utils::ConditionIterator;

pub fn make_validity_constraint<'ctx>(packages: &Vec<Package>,
                                    package_variables: &Vec<Vec<Ast<'ctx>>>) -> Option<Ast<'ctx>> {
    package_variables.iter()
                     .map(|state| {
                         packages.iter()
                                 .zip(state.iter())
                                 .filter_map(move |(package, package_var)| {
                                     package_validity_constraint(package, package_var, &state)
                                 })
                     })
                     .flatten()
                     .conjunction()
}

fn package_validity_constraint<'ctx>(package: &Package,
                                     package_variable: &Ast<'ctx>,
                                     state: &Vec<Ast<'ctx>>) -> Option<Ast<'ctx>> {

    let all_dependencies_installed: Option<Ast> =
        package.dependencies.iter()
                            .filter_map(|dep| dependency_constraint(package_variable, dep, state))
                            .conjunction();

    let no_conflicting_installed: Option<Ast> =
        package.conflicts.iter()
                         .filter_map(|conflict| conflict_constraint(package_variable, conflict, state))
                         .conjunction();

    match (all_dependencies_installed, no_conflicting_installed) {
        (Some(deps), Some(conf)) => Some(deps.and(&[&conf])),
        (Some(deps), None) => Some(deps),
        (None, Some(conf)) => Some(conf),
        (None, None) => None
    }
}

fn dependency_constraint<'ctx>(package_variable: &Ast<'ctx>,
                               alternatives: &Vec<PackageConstraint>,
                               state: &Vec<Ast<'ctx>>) -> Option<Ast<'ctx>> {
    let dependencies_installed = alternatives.iter()
                                             .map(|alternative| {
                                                 alternative.possibilities()
                                                            .iter()
                                                            .map(|&package_id| state[package_id].clone())
                                             })
                                             .flatten()
                                             .disjunction();
    dependencies_installed.map(|dep| package_variable.implies(&dep))
}

fn conflict_constraint<'ctx>(package_variable: &Ast<'ctx>,
                             constraint: &PackageConstraint,
                             state: &Vec<Ast<'ctx>>) -> Option<Ast<'ctx>> {
    let possibilities = constraint.possibilities();
    if !possibilities.is_empty() {
        possibilities.iter()
                     .map(|&package_id| state[package_id].clone())
                     .disjunction()
                     .map(|any_conflicts_installed| {
                          package_variable.not().and(&[&any_conflicts_installed])
                     })
    } else {
        None
    }
}
